jdaviz/utils.py

```diff
@@ -260,3 +262,32 @@ def __call__(self):
 
     def reset(self):
         self.counter = -1
+
+
+def get_subset_type(subset):
+    """
+    Determine the subset type of a subset or layer
+
+    Parameters
+    ----------
+    subset : should have ``subset_state`` as an attribute, otherwise will return ``None``.
+
+    Returns
+    -------
+    subset_type : str or None
+        'spatial', 'spectral', or None
+    """
+    if not hasattr(subset, 'subset_state'):
+        return None
+
+    while hasattr(subset.subset_state, 'state1'):
+        # this assumes no mixing between spatial and spectral subsets and just
+        # taking the first component (down the hierarchical tree) to determine the type
+        subset = subset.subset_state.state1
+
+    if isinstance(subset.subset_state, RoiSubsetState):
+        return 'spatial'
+    elif isinstance(subset.subset_state, RangeSubsetState):
+        return 'spectral'
+    else:
+        return None
```

Comment:
Maybe this isn't correct, but the point is you want to break out of the loop as soon as you find a match, right?
```suggestion
        if isinstance(subset.subset_state, RoiSubsetState):
            return 'spatial'
        elif isinstance(subset.subset_state, RangeSubsetState):
            return 'spectral'
        else:
            return None
```

Also, what if what you are looking for is in `state2`? Can't we use Jesse's tree traversal here?